### ISSUE TITLE
RELEASE OF PACKAGE VERSION 1.5.0 IS A DISASTROUS MALFUNCTION

### ISSUE THREAD

**Issue by AlseinX** [gakonst/ethers-rs#2224]
The crate resolving system uses [**Semantic Versioning**](https://semver.org/#semantic-versioning-200), which guarantees that a minor version must not introduce any breaking change once it reaches `1.0.0`.
Thus, the version resolving behavior is that, as long as there is a newer minor version than the specified version, the newer one is automatically selected. Although a minor version could be enforced by [Tilde or Equals requirement](https://doc.rust-lang.org/cargo/reference/resolver.html#semver-compatibility), it only works for direct dependencies.
Since `ethers` crate is an all-in-one entry of the detailed packages, these packages are all resolved as indirect dependencies that could not be enforced to an exact minor version, the release of `1.5.0` makes `1.0.2` no longer accessible through `ethers`. The dependents have to depend on `ethers-*` packages separately, which could be specified to an exact version but is obviously unexpected.
The release of `1.5.0` is disastrously breaking the existing code from compiling, which is a shocking fault to the dependents.

**Comment by Kitryn** [comment]
just chiming in with the same issue -- fully agree that breaking changes should not have been pushed under a minor version increment. This was breaking all of our builds containing third party dependencies which pulled in 1.5.0 under semver (we already pin to 1.0.2 for our internal crates), so we ended up having to write a `[patch.crates-io]` in our Cargo.toml to override dependencies to `ethers = { git = "https://github.com/gakonst/ethers-rs.git", tag = "1.0.2" }`

**Comment by nabeel99** [comment]
@Kitryn  hey, ran into the same issue, were you able to find a temporary fix for this ? applying your above solution dent work for me, been getting abigen errors as it defaults to using 1.5.0 version of it.

**Comment by Kitryn** [comment]
@nabeel99 For our use-case we depend on the subpackages individually, so we ended up just pinning all of them to the git tag:

```
ethers-addressbook = { git = "https://github.com/gakonst/ethers-rs.git", tag = "1.0.2" }
ethers-contract = { git = "https://github.com/gakonst/ethers-rs.git", tag = "1.0.2" }
ethers-core = { git = "https://github.com/gakonst/ethers-rs.git", tag = "1.0.2" }
ethers-etherscan = { git = "https://github.com/gakonst/ethers-rs.git", tag = "1.0.2" }
ethers-middleware = { git = "https://github.com/gakonst/ethers-rs.git", tag = "1.0.2" }
ethers-providers = { git = "https://github.com/gakonst/ethers-rs.git", tag = "1.0.2", features = ["rustls"] }
ethers-signers = { git = "https://github.com/gakonst/ethers-rs.git", tag = "1.0.2" }
ethers-contract-abigen = { git = "https://github.com/gakonst/ethers-rs.git", tag = "1.0.2" }
ethers-solc = { git = "https://github.com/gakonst/ethers-rs.git", tag = "1.0.2" }

...
...

[patch.crates-io]
ethers = { git = "https://github.com/gakonst/ethers-rs.git", tag = "1.0.2" }
ethers-contract-derive = { git = "https://github.com/gakonst/ethers-rs.git", tag = "1.0.2" }
```

the first section is what our internal code depends on, and the subsequent patch section was sufficient for us to patch whatever 3rd party dependencies we were using (i.e. ethers-flashbots - though I see that they've updated for 1.5.0 already)

Note that there's some chatter in the ethers-rs telegram group about yanking 1.5.0 and re-releasing it as 2.0, so the above might be moot in a few days

**Comment by nabeel99** [comment]
Yup reading the cargo book, the patch thing is the way to gi
An alternative is to force the exact version using crate = "= <version>" syntax and running cargo update -p <package_name> to update the catgo lock.

**Comment by Kitryn** [comment]
yep, before this whole thing we specified our direct ethers dependencies using `~1.0.2` (which translates to >=1.0.2, <1.1) so were not directly affected, however our third party dependencies pulled in 1.5.0 which broke them; patching them required the `[patch.crates.io]` section, which then broke our internal packages because cargo started failing builds with `note: perhaps two different versions of crate ethers_signers are being used?`, thus we had to also redirect our direct dependency on ethers to the git repo as well

a whole mess, really

**Comment by gakonst** [comment]
Apologies. This was my bad, I rushed the release. Plan is to yank and republish as 2.0 by end of weekend. Appreciate the patience.

**Comment by gakonst** [comment]
This is fixed as 1.5 is yanked.
